genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/FileSystemFetchingCacheService.java

```diff
@@ -0,0 +1,343 @@
+ * this file and then downloads the resource to a data.tmp file. Once the download is complete
+ * data.tmp is atomically moved to data.
+ * For e.g for a hash value of 6d331abc92bc8244bc5d41e2107f303a and last modified = 1525456404
+ * cache entries would like the following
+ * {base_dir}/6d331abc92bc8244bc5d41e2107f303a/1525456404/data
+ * {base_dir}/6d331abc92bc8244bc5d41e2107f303a/1525456404/lock
+ * <p>
+ * Deletion of older versions
+ *
+ * @author standon
+ * @since 4.0.0
+ */
+@Service
+@Lazy
+@Slf4j
+@Component
+class FileSystemFetchingCacheService implements FetchingCacheService {
+
+    public static final String LOCK_FILE_NAME = "lock";
+    public static final String DATA_FILE_NAME = "data";
+    public static final String DATA_DOWNLOAD_FILE_NAME = "data.tmp";
+    protected final ResourceLoader resourceLoader;
+    protected final File cacheDirectory;
+    protected final LockingFactory lockingFactory;
+    private final ExecutorService oldResourcesCleanUpService = Executors.newFixedThreadPool(1);
+
+
+    FileSystemFetchingCacheService(
+        final ResourceLoader resourceLoader,
+        final ArgumentDelegates.CacheArguments cacheArguments,
+        final LockingFactory lockingFactory
+    ) {
+        this.resourceLoader = resourceLoader;
+
+        this.cacheDirectory = cacheArguments.getCacheDirectory();
+
+        this.lockingFactory = lockingFactory;
+
+        createDirectoryStructureIfNotExists(cacheDirectory);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void get(final URI sourceFileUri, final File destinationFile) throws DownloadException, IOException {
+        lookupOrDownload(sourceFileUri, destinationFile);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void get(final Set<Pair<URI, File>> sourceDestinationPairs) throws DownloadException, IOException {
+        for (final Pair<URI, File> sourceDestinationPair : sourceDestinationPairs) {
+            get(sourceDestinationPair.getKey(), sourceDestinationPair.getValue());
+        }
+    }
+
+    protected void lookupOrDownload(
+        final URI sourceFileUri,
+        final File destinationFile
+    ) throws DownloadException, IOException {
+
+        final String uriString = sourceFileUri.toASCIIString();
+
+        log.info("Lookup: {}", uriString);
+
+        // Unique id to store the resource on local disk
+        final String resourceCacheId = getResourceCacheId(sourceFileUri);
+
+        // Get a handle to the resource
+        final Resource resource = resourceLoader.getResource(uriString);
+
+        if (!resource.exists()) {
+            throw new DownloadException("Resource not found: " + uriString);
+        }
+
+        final long resourceLastModified = resource.lastModified();
+
+        //Create the resource version dir in cache if it does not exist
+        final File cacheResourceVersionDir = getCacheResourceVersionDir(
+            resourceCacheId,
+            resourceLastModified
```

Comment:
I did not do it because then we have to worry about the granularity. For e.g if we chose the format as YYYYMMDD, then two version changes int the same day will convert to the same value. This way we dont worry about granularity

Reply:
A timestamp with format `YYYYMMDD` would indeed bad for this (other than being hard to read). But date formats can be customized...

Reply:
`resource.lastModified` documentation says -`Determine the last-modified timestamp for this resource.` I will assume it will be the unix timestamp and parse it to only seconds in that case.
https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/core/io/Resource.html#lastModified--

Reply:
Typo - will parse it to millis. format - yyyy_MM_dd_HH_mm_ss_SSS